lib/spack/docs/tutorial_buildsystems.rst

```diff
@@ -0,0 +1,351 @@
+    ==> Found 4 versions of callpath:
+
+    1.0.3  https://github.com/LLNL/callpath/archive/v1.0.3.tar.gz
+    1.0.2  https://github.com/LLNL/callpath/archive/v1.0.2.tar.gz
+    1.0.1  https://github.com/LLNL/callpath/archive/v1.0.1.tar.gz
+    1.0    https://github.com/LLNL/callpath/archive/v1.0.tar.gz
+
+    ==> How many would you like to checksum? (default is 1, q to abort) 1
+    ==> Downloading...
+    ==> Fetching https://github.com/LLNL/callpath/archive/v1.0.3.tar.gz
+    ######################################################################## 100.0%
+    ==> Checksummed 1 version of callpath
+    ==> This package looks like it uses the cmake build system
+    ==> Created template for callpath package
+    ==> Created package file: /Users/mamelara/spack/var/spack/repos/builtin/packages/callpath/package.py
+
+
+which then produces the following template:
+
+.. literalinclude:: tutorial/examples/Cmake/0.package.py
+   :language: python
+
+Again we fill in the details:
+
+.. literalinclude:: tutorial/examples/Cmake/1.package.py
+   :language: python
+
+Spack will automatically add -DCMAKE_INSTALL_PREFIX, -DCMAKE_BUILD_TYPE, 
+-DCMAKE_VERBOSE_MAKEFILE as well as a number of other options to cmake without
+the user having to explictly add those arguments. 
+
+In callpath, we want to add options to CALLPATH_WALKER as well as add compiler flags.
+We add the following options like so:
+
+.. literalinclude:: tutorial/examples/Cmake/2.package.py
+   :language: python
+
+Now we can control our build options using cmake_args. Again if defaults are
+sufficient enough for the package, we can leave out cmake_args in our package.py
+
+CMake classes allow for control of other features in the build system. For example,
+you can specify the path to the "out of source" build directory and also point
+to the root of the CMakeLists.txt file if it is placed in a non-standard 
+location.
+
+--------------
+PythonPackage
+--------------
+
+Python extensions and modules are built differently from source than most
+applications. Python uses a setup.py script to install Python modules. 
+The script consists of a call to setup() which provides the information 
+required to building a module to Distutils. If you're familiar with pip or 
+easy_install, setup.py does the same thing.
+
+These modules are usually installed using the following line:
+
+.. code-block:: console
+    
+    $ python setup.py install <module>
+
+There are also a list of commands and phases that you can call. To see the full
+list you can run:
+    
+.. code-block:: console
+    
+    $ python setup.py --help-commands
+
+For each of these commands, Spack has a method that can be overridden. 
+
+Using the Package class is possible but it brings with it a lot of unnecessary
+methods that are useless for a Python package. Instead, Spack has a PythonPackage
+subclass that allows packagers of Python modules to be able to invoke setup.py
+and use Distutils.  
```

Comment:
maybe note here that this makes the PythonPackage based class look much more familiar to a typical python user.